06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+# Kafka Roller 2.0
+
+## Current situation
+
+The Kafka Roller is an internal Cluster Operator component that's responsible for coordinating the rolling restart or reconfiguration of Kafka pods when:
+- non-dynamic reconfigurations needs to be applied
+- update in Kafka CRD is detected
+- a certificate is renewed
+- pods have been manually annotated by the user for controlled restarts
+- pod is stuck and is out of date
+- Kafka broker is unresponsive to Kafka Admin connections
+
+A pod is considered stuck if it is in one of following states:
+- `CrashLoopBackOff`
+- `ImagePullBackOff`
+- `ContainerCreating`
+- `Pending` and `Unschedulable`
+
+### Known Issues
+
+The existing KafkaRoller suffers from the following shortcomings:
+- Although it is safe and straightforward to restart one broker at a time, this process is slow in large clusters ([related issue](https://github.com/strimzi/strimzi-kafka-operator/issues/8547)).
+- It does not account for partition preferred leadership. As a result, there may be more leadership changes than necessary during a rolling restart, consequently impacting tail latency.
+- Hard to reason about when things go wrong. The code is complex to understand and it's not easy to determine why a pod was restarted from logs that tend to be noisy.
+- Potential race condition between Cruise Control rebalance and KafkaRoller that could cause partitions under minimum in sync replica. This issue is described in more detail in the `Future Improvements` section.
+- The current code for KafkaRoller does not easily allow growth and adding new functionality due to its complexity.
+
+
+The following non-trivial fixes and changes are missing from the current KafkaRoller's KRaft implementation:
+- Currently KafkaRoller has to connect to brokers successfully in order to get KRaft quorum information and determine whether a controller node can be restarted. This is because it was not possible to directly talk to KRaft controllers at the time before [KIP 919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum+and+add+Controller+Registration) was implemented. The issue is raised [here](https://github.com/strimzi/strimzi-kafka-operator/issues/9692).
```

Comment:
Yes. At the time of writing, KIP-919 was implemented but not used in Strimzi yet so the idea was to implement this in the new roller, but now that's blocked behind another upstream [issue](https://issues.apache.org/jira/browse/KAFKA-16781). I can remove this section, as it's now missing because of upstream.